Samples/GLTFViewer/src/GLTFViewer.cpp

```diff
@@ -389,67 +406,27 @@ void GLTFViewer::UpdateUI()
 
 void GLTFViewer::CreateEnvMapPSO()
 {
-    ShaderCreateInfo                               ShaderCI;
-    RefCntAutoPtr<IShaderSourceInputStreamFactory> pShaderSourceFactory;
-    m_pEngineFactory->CreateDefaultShaderSourceStreamFactory("shaders", &pShaderSourceFactory);
-    ShaderCI.pShaderSourceStreamFactory = pShaderSourceFactory;
-    ShaderCI.SourceLanguage             = SHADER_SOURCE_LANGUAGE_HLSL;
-    ShaderCI.UseCombinedTextureSamplers = true;
-
-    ShaderMacroHelper Macros;
-    Macros.AddShaderMacro("TONE_MAPPING_MODE", "TONE_MAPPING_MODE_UNCHARTED2");
-    ShaderCI.Macros = Macros;
-
-    ShaderCI.Desc.ShaderType = SHADER_TYPE_VERTEX;
-    ShaderCI.Desc.Name       = "Environment map VS";
-    ShaderCI.EntryPoint      = "main";
-    ShaderCI.FilePath        = "env_map.vsh";
-    RefCntAutoPtr<IShader> pVS;
-    m_pDevice->CreateShader(ShaderCI, &pVS);
-
-    ShaderCI.Desc.Name       = "Environment map PS";
-    ShaderCI.EntryPoint      = "main";
-    ShaderCI.FilePath        = "env_map.psh";
-    ShaderCI.Desc.ShaderType = SHADER_TYPE_PIXEL;
-    RefCntAutoPtr<IShader> pPS;
-    m_pDevice->CreateShader(ShaderCI, &pPS);
-
-    GraphicsPipelineStateCreateInfo PSOCreateInfo;
-    PipelineStateDesc&              PSODesc          = PSOCreateInfo.PSODesc;
-    GraphicsPipelineDesc&           GraphicsPipeline = PSOCreateInfo.GraphicsPipeline;
-
-    PSODesc.ResourceLayout.DefaultVariableType = SHADER_RESOURCE_VARIABLE_TYPE_MUTABLE;
-
-    // clang-format off
-    ImmutableSamplerDesc ImmutableSamplers[] =
-    {
-        {SHADER_TYPE_PIXEL, "EnvMap", Sam_LinearClamp}
-    };
-    // clang-format on
-    PSODesc.ResourceLayout.ImmutableSamplers    = ImmutableSamplers;
-    PSODesc.ResourceLayout.NumImmutableSamplers = _countof(ImmutableSamplers);
-
-    // clang-format off
-    ShaderResourceVariableDesc Vars[] = 
+    struct GraphicsPipelineInfo
     {
-        {SHADER_TYPE_PIXEL, "cbCameraAttribs",       SHADER_RESOURCE_VARIABLE_TYPE_STATIC},
-        {SHADER_TYPE_PIXEL, "cbEnvMapRenderAttribs", SHADER_RESOURCE_VARIABLE_TYPE_STATIC}
+        TEXTURE_FORMAT RTVFmt;
+        TEXTURE_FORMAT DSVFmt;
+    } UserData{m_pSwapChain->GetDesc().ColorBufferFormat, m_pSwapChain->GetDesc().DepthBufferFormat};
+
+    auto Modify = [](PipelineStateCreateInfo* pPipelineCI_, void* pUserData_) {
+        auto* pUserData                                = static_cast<GraphicsPipelineInfo*>(pUserData_);
+        auto* pPipelineCI                              = static_cast<GraphicsPipelineStateCreateInfo*>(pPipelineCI_);
```

Comment:
There is really no need to use the user data here. You can capture the environment, but you need to assign lamda to the std::function